iroh: Split into multiple packages

matthiasbeyer opened the pull request:
Closes #454227

Might not be solved the way nixpkgs wants these problems being solved, I don't know.


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

CC @cameronfyfe

matthiasbeyer committed:
iroh: Split into multiple packages

Signed-off-by: Matthias Beyer <mail@beyermatthias.de>

matthiasbeyer committed:
iroh: Add myself as maintainer

Signed-off-by: Matthias Beyer <mail@beyermatthias.de>

eclairevoyant reviewed (changes requested):
That definitely won't work, the whole point of `by-name` is that the directory name corresponds to the attribute name.

You can move it out of `by-name` and use the following pattern: https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md#recommendation-for-new-packages-with-multiple-versions

matthiasbeyer committed:
squash! iroh: Split into multiple packages

Because by-name is not applicable anymore here, we move the package to
tools/networking instead.

eclairevoyant commented on pkgs/tools/networking/iroh/iroh-relay.nix:
```diff
@@ -0,0 +1,12 @@
+{
+  callPackage,
+  ...
```
```suggestion
```

matthiasbeyer replied:
Why?

eclairevoyant commented on pkgs/tools/networking/iroh/iroh-dns-server.nix:
```diff
@@ -0,0 +1,11 @@
+{
+  callPackage,
+  ...
```
```suggestion
```

eclairevoyant commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5065,6 +5065,9 @@ with pkgs;
   };
 
+  iroh-relay = callPackage ../tools/networking/iroh/iroh-relay.nix { };
+  iroh-dns-server = callPackage ../tools/networking/iroh/iroh-dns-server.nix { };
```
This needs to follow the code pattern I linked above, or the `nixpkgs-vet` check will not pass.

```suggestion
  inherit ({
    iroh-relay = callPackage ../tools/networking/iroh/iroh-relay.nix { };
    iroh-dns-server = callPackage ../tools/networking/iroh/iroh-dns-server.nix { };
  })
  iroh-relay
  iroh-dns-server;
```

matthiasbeyer replied:
What? I never seen this pattern... Is there any rationale why I would use this?

eclairevoyant reviewed (changes requested):
We also need a `throw` alias for `iroh` since it no longer exists as an attribute.

To avoid CI failures, please ensure to run `nix-shell --run treefmt` after making the changes.

matthiasbeyer commented on pkgs/tools/networking/iroh/irohPackage.nix:
```diff
@@ -20,14 +25,27 @@ rustPlatform.buildRustPackage rec {
+    [
+      featureFlag
+      binaryFlag
+    ];
```
Much less readable IMO

eclairevoyant replied:
This is a standard pattern in nixpkgs. But either way you still have to fix the two points I mentioned (remove parens, use the list structure). If you want to make it verbose and frankly less readable with the `let`-`in` structure, that's your business.

cameronfyfe commented:
I'll try to pick this up when I have some time, it's been on my mental todo list to fix/update this package since the `iroh` binary was removed from this crate.
